### DISCUSSION TITLE
Share a instance across the API life cycle in FastAPI without passing it every class/function (Scope instance)

### DISCUSSION THREAD

**Discussion by hasansustcse13** [fastapi/fastapi#10699] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class HelperClass:
    def __init__(self):
        self.token_count = 0

    def increase(self):
        self.token_count += 1


class ClassB:
    def __init__(self, helper):
        self.helper = helper

    def do_something(self):
        print("ClassB is doing something.")
        self.helper.increase()


class ClassC:
    def __init__(self, helper):
        self.helper = helper

    def perform_task(self):
        print("ClassC is performing a task.")
        self.helper.increase()


class ChatService:
    def __init__(self):
        helper = HelperClass()
        self.class_b = ClassB(helper)
        self.class_c = ClassC(helper)

    def execute(self):
        print("ClassA is executing.")
        self.class_b.do_something()
        self.class_c.perform_task()
        for i in range(100):
            yield i
            time.sleep(.5)

@router.post('/chat')
async def chat_async():
    return StreamingResponse(ChatService().execute(), media_type='text/event-stream')
```

### Description

I want to share a class instance across the API call. Let's say I have class named HelperClass. I want to share it across the all code. In the example, the class needs in some other class. I don't want to pass it all classes. I just want to use it only where needed. I want similar kind of functionality (Scoped DI) of .NET API. How do I do this in FastAPI? I know it may possible to achieve same functionality using @contextmanager decorator. But it won't work with StreamingResponse.

Additional Question: Does FastAPI depends work outside of router function? 

### Operating System

Windows

### FastAPI Version

fastapi==0.104.1

### Pydantic Version

pydantic_core==2.10.1

### Python Version

3.11

**Comment by bhushanasati25** [comment]
Hi! In FastAPI I would separate this into two cases:

1. App-wide shared object
2. Per-request / per-API-call object

For your example, it sounds like you want the second one: one `HelperClass` instance for one `/chat` request, shared by the classes used during that request.

FastAPI’s `Depends` can create that per-request object, but it only runs inside FastAPI’s dependency graph. So it works in path operations and other dependencies. It does not automatically inject things into arbitrary classes/methods outside that graph like a .NET DI container would.

A simple version would be:

```python
from typing import Annotated
from fastapi import Depends
from fastapi.responses import StreamingResponse

class HelperClass:
    def __init__(self):
        self.token_count = 0

    def increase(self):
        self.token_count += 1


class ClassB:
    def __init__(self, helper: HelperClass):
        self.helper = helper

    def do_something(self):
        self.helper.increase()


class ClassC:
    def __init__(self, helper: HelperClass):
        self.helper = helper

    def perform_task(self):
        self.helper.increase()


class ChatService:
    def __init__(self, helper: HelperClass):
        self.class_b = ClassB(helper)
        self.class_c = ClassC(helper)

    def execute(self):
        self.class_b.do_something()
        self.class_c.perform_task()

        for i in range(100):
            yield str(i)


def get_helper():
    return HelperClass()


def get_chat_service(
    helper: Annotated[HelperClass, Depends(get_helper)],
):
    return ChatService(helper)


@router.post("/chat")
async def chat_async(
    chat_service: Annotated[ChatService, Depends(get_chat_service)],
):
    return StreamingResponse(
        chat_service.execute(),
        media_type="text/event-stream",
    )
```

This gives you one `HelperClass` instance for that request, and both `ClassB` and `ClassC` share it.

If the object should live for the whole application lifetime instead, create it during app startup/lifespan and store it on `app.state`, then access it through `request.app.state`. But for request-scoped state, I would avoid a global singleton because concurrent requests would share the same object.

About your extra question: `Depends` does not really “work” by itself outside a route/dependency context. If you write something like this inside a normal class:

```python
def some_method(helper = Depends(get_helper)):
    ...
```

FastAPI will not resolve it unless that function is being called as part of the FastAPI dependency system. Outside that system, pass the dependency explicitly, use a small factory function, or use an external DI library.

One note for `StreamingResponse`: if the helper/resource must stay alive while streaming, make sure its lifetime covers the generator. With newer FastAPI versions, request-scoped `yield` dependencies are designed to close after the response is sent, but if you are on an older FastAPI version and run into cleanup happening too early, put the setup/cleanup inside the streaming generator itself or upgrade FastAPI.
